README.md
# react-native-pin-code

![last release](https://badgen.net/npm/v/react-native-pin-code) ![monthly download](https://badgen.net/npm/dm/react-native-pin-code) ![license](https://badgen.net/github/license/gkueny/react-native-pin-code)

A simple pin code component

![example with blur background](https://media.giphy.com/media/xUPGcffB0VeaMd6DSM/giphy.gif)
![a simple example](https://media.giphy.com/media/3oKIPsotgoJ8ZGEr5u/giphy.gif)

## Installation

npm install react-native-pin-code --save

## Example

```js
<CodePin
  code="2018" // code.length is used if you not pass number prop
  success={() => console.log('hurray!')} // If user fill '2018', success is called
  text="A simple Pin code component" // My title
  error="You fail" // If user fail (fill '2017' for instance)
  autoFocusFirst={false} // disabling auto-focus
/>
```

or

```js
<CodePin
  number={4} // You must pass number prop, it will be used to display 4 (here) inputs
  checkPinCode={(code, callback) => callback(code === '1234')}
  // Check manually code (ask server for instance)
  // and call callback function with
  //    true  (code pin is correct)
  // or false (code pin is false)
  success={() => console.log('hurray!')} // If user fill '2018', success is called
  text="A simple Pin code component" // My title
  error="You fail" // If user fail (fill '2017' for instance)
  autoFocusFirst={false} // disabling auto-focus
/>
```

> Code prop is not needed if checkPinCode is used

## props

| prop              | type                      | description                                                                                                                                                                                                      | isRequired   | default value                                                                                                                                                                                                      |
| ----------------- | ------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| code              | number                    | pin code                                                                                                                                                                                                         | not required | ''                                                                                                                                                                                                                 |
| success           | function                  | call on success                                                                                                                                                                                                  | required     |                                                                                                                                                                                                                    |
| checkPinCode      | function                  | this function is called to check pin code. The parameters are : `code` (the code filled by user) and a `callback` function. callback must be called with true (code pin is correct) or false (code pin is wrong) | not required | null                                                                                                                                                                                                               |
| obfuscation       | boolean                   | If true, obfuscate code with '\*'                                                                                                                                                                                |              | false                                                                                                                                                                                                              |
| text              | string                    | text to display as title                                                                                                                                                                                         |              | 'Pin code.'                                                                                                                                                                                                        |
| error             | string                    | text to display on error                                                                                                                                                                                         |              | 'Bad pin code.'                                                                                                                                                                                                    |
| number            | number                    | number of input to display                                                                                                                                                                                       |              | 4                                                                                                                                                                                                                  |
| autoFocusFirst    | boolean                   | auto focus first input                                                                                                                                                                                           |              | true                                                                                                                                                                                                               |
| containerStyle    | object, StyleSheet object | container style                                                                                                                                                                                                  |              | { height: 150, width: width - 30, backgroundColor : '#FFF' }                                                                                                                                                       |
| containerPinStyle | object, StyleSheet object | inputs container style                                                                                                                                                                                           |              | { width: width - 30, height: 40, flexDirection: 'row', justifyContent: 'space-around', alignItems: 'center', marginTop: 20 }                                                                                       |
| pinStyle          | object, StyleSheet object | input style                                                                                                                                                                                                      |              | { backgroundColor : '#F0F0F0', textAlign: 'center', flex: 1, marginLeft: 20, marginRight: 20, borderRadius: 5, shadowColor: '#000000', shadowOffset: {width: 1,height : 1}, shadowRadius: 5, shadowOpacity : 0.4 } |
| textStyle         | object, StyleSheet object | text style                                                                                                                                                                                                       |              | { textAlign: 'center', color: 'gray', fontSize: 20, marginTop: 30 }                                                                                                                                                |
| errorStyle        | object, StyleSheet object | error text style                                                                                                                                                                                                 |              | { textAlign: 'center', color: 'red', paddingTop: 10 }                                                                                                                                                              |

All other `props` are pass down to `TextInput` component.

For instance you can customize keyboardtype :

```js
<CodePin
  //...
  keyboardType="numeric"
/>
```

## Functions

To use pin code component function, use ref :

```js
  <CodePin
    ref={ref => this.ref = ref}
    ...
  />
```

## clean

clean inputs and focus first input.

```js
this.ref.clean();
```

## focus

focus input at specified index

```js
this.ref.focus(1);
```

## Examples

> [react-native-pin-code-examples](https://github.com/gkueny/react-native-pin-code-examples)

To test examples (use [expo](https://expo.io)) :

```
  git clone https://github.com/gkueny/react-native-pin-code-examples
  cd blurExample
  npm install
  npm run ios
```

or

```
  git clone https://github.com/gkueny/react-native-pin-code-examples
  cd simpleExample
  npm install
  npm run ios
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [v0.5.3] - 2018-10-01

### Fixed

- Fix clean inputs (see [#22](https://github.com/gkueny/react-native-pin-code/issues/22))

## [v0.5.1] - 2018-09-09

### Fixed

- Fix clear input fields on success (see [#16](https://github.com/gkueny/react-native-pin-code/issues/16))
- Fix focus

## [v0.5.0] - 2018-09-03

### Added

- Supports auto-fill and past of codes.

### Fixed

- Fixed crash which occurs when using iOS 12's security code auto-fill feature.
- Corrected Typos in Readme/Changelog

## [v0.4.1] - 2018-02-09

### Fixed

- allow to not pass `code` prop.

## [v0.4.0] - 2018-02-09

### Added

- handle backspace event

## [v0.3.0] - 2018-02-04

### Added

- New prop `checkPinCode`(optional). Called to check pin code (`code` prop is not used if `checkPinCode` prop is present)
- New prop `obfuscation` (optional). Obfuscate code if set to true

### Changed

- Update Readme to add simple example.

### Fixed

- Use `code.length` if `number` prop is not used.


Examples/blurExample/index.ios.js
import React, { Component } from 'react';
import {
  KeyboardAvoidingView,
  AppRegistry,
  StyleSheet,
  Dimensions,
  Text,
  View
} from 'react-native';

import CodePin from './pin-code/pin-code';
const { BlurView } = require('react-native-blur');

const { height, width } = Dimensions.get('window');

export default class blurExample extends Component {
  constructor() {
    super();

    this.state = {
      displayCodePin: true,
      success: ''
    };
  }

  onSuccess = () => {
    this.ref.clean();

    this.setState({
      displayCodePin: false,
      success: 'Okeyyyyy !'
    });
  };

  render() {
    return (
      <View style={styles.container}>

        <Text style={styles.success}>
          {this.state.success}
        </Text>

        {this.state.displayCodePin
          ? <BlurView blurType="dark" blurAmount={10} style={styles.blur}>

              <KeyboardAvoidingView
                behavior={'position'}
                contentContainerStyle={styles.avoidingView}
              >
                <CodePin
                  ref={ref => (this.ref = ref)}
                  code="290317"
                  number={6}
                  success={this.onSuccess}
                  containerStyle={styles.containerCodePin}
                  pinStyle={styles.pinStyle}
                  textStyle={{ fontSize: 12 }}
                  text={'Protected area'}
                  errorStyle={{ fontSize: 10 }}
                  error={'Look at the code ;)'}
                  keyboardType="numeric"
                />
              </KeyboardAvoidingView>

            </BlurView>
          : null}

      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF'
  },
  blur: {
    position: 'absolute',
    justifyContent: 'center',
    alignItems: 'center',
    top: 0,
    left: 0,
    width: width,
    height: height
  },
  avoidingView: {
    borderRadius: 10,
    height: 150,
    width: width - 30
  },
  containerCodePin: {
    borderRadius: 10
  },
  pinStyle: {
    marginLeft: 5,
    marginRight: 5
  },
  success: {
    fontSize: 20,
    color: 'green',
    textAlign: 'center'
  }
});

AppRegistry.registerComponent('blurExample', () => blurExample);
